source/common/stats/symbol_table_impl.cc

```diff
@@ -75,11 +174,87 @@ Symbol SymbolTableImpl::toSymbol(absl::string_view sv) {
   return result;
 }
 
-absl::string_view SymbolTableImpl::fromSymbol(const Symbol symbol) const {
+absl::string_view SymbolTable::fromSymbol(const Symbol symbol) const
+    EXCLUSIVE_LOCKS_REQUIRED(lock_) {
   auto search = decode_map_.find(symbol);
   ASSERT(search != decode_map_.end());
   return search->second;
 }
 
+void SymbolTable::newSymbol() EXCLUSIVE_LOCKS_REQUIRED(lock_) {
+  if (pool_.empty()) {
+    next_symbol_ = ++monotonic_counter_;
+  } else {
+    next_symbol_ = pool_.top();
+    pool_.pop();
+  }
+  // This should catch integer overflow for the new symbol.
+  ASSERT(monotonic_counter_ != 0);
+}
+
+bool SymbolTable::lessThan(const StatName& a, const StatName& b) const {
+  // Constructing two temp vectors during lessThan is not strictly necessary.
+  // If this becomes a performance bottleneck (e.g. during sorting), we could
+  // provide an iterator-like interface for incrementally decoding the symbols
+  // without allocating memory.
+  SymbolVec av = SymbolEncoding::decodeSymbols(a.data(), a.numBytes());
+  SymbolVec bv = SymbolEncoding::decodeSymbols(b.data(), b.numBytes());
+  for (size_t i = 0, n = std::min(av.size(), bv.size()); i < n; ++i) {
+    if (av[i] != bv[i]) {
+      Thread::LockGuard lock(lock_);
+      return fromSymbol(av[i]) < fromSymbol(bv[i]);
+    }
+  }
+  return av.size() < bv.size();
+}
+
+StatNameStorage::StatNameStorage(absl::string_view name, SymbolTable& table) {
+  SymbolEncoding encoding = table.encode(name);
+  bytes_.reset(new uint8_t[encoding.bytesRequired()]);
+  encoding.moveToStorage(bytes_.get());
+}
+
+StatNameStorage::~StatNameStorage() {
+  // StatNameStorage is not fully RAII: you must call free(SymbolTable&) to
```

Comment:
Sorry how does this save 8 bytes?

Reply:
If we want to make this RAII we need to store a reference to the SymbolTable& which would require 8 bytes overhead per StatNameStorage instance. Added comment.